rfcs/47-better-error-handing.md

```diff
@@ -0,0 +1,209 @@
+    	- defined in [go-storage/services/error.go]
+    	- have `Err` and context fields, don't have `Op`
+    	- returned by constructor methods, where `Err` is set to a sentinel error, which can be viewed as the error `struct`s' classification
+
+![](https://user-images.githubusercontent.com/37948597/116873067-9ce86c80-ac49-11eb-9409-91dcc84da340.png)
+### Problems
+
+#### Unexpected Errors: Abstraction Leak
+
+Wrapping (partial) SDK errors will lead to abstraction leak. In terms of error handling, being vendor agnostic means we should provide a unified custom error to hide the SDK errors, e.g., we convert some SDK errors into our `ErrObjectNotExist`. However, unexpected errors will be wrapped directly. It is possible that users use `errors.Is` & `errors.As` targeting SDK errors. Then problems will come when we add a new kind of expected error and convert some SDK errors into it. It may be confusing that some SDK errors are wrapped, while the others are not. Below is an example of such partial conversion.
+
+```go
+func formatError(err error) error {
+	e, ok := err.(awserr.RequestFailure)
+	if !ok {
+		return err
+	}
+
+	switch e.Code() {
+	// AWS SDK will use status code to generate awserr.Error, so "NotFound" should also be supported.
+	case "NoSuchKey", "NotFound":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)
+	case "AccessDenied":
+		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, err)
+	// What if we add a new case, but users already handle it themselves?
+	}
+
+	return err
+}
+```
+
+Although we declared in `11-error-handling` that unexpected errors could be changed or disappeared while dependence upgraded and no changelog for them, we'd better simply eliminate the possibility of using them.
+
+#### Expected Errors
+
+`11-error-handling` is not detailed enough. It does not specify how to define and use expected errors. Current practice has too many cases, and has some problems. Both users and implementors may be confused which kind of error should be returned at which place.
+
+- Ad-hoc string errors are not user-friendly. The user cannot use `errors.Is` & `errors.As` to handle such errors specially. We should avoid using them.
+
+- Another layer of error `struct`s may be confusing.
+  - It may be mistaken for top-level errors.
+  - Their `Err` fields are also confusing.
+    - Will they wrap yet another layer of error `struct`s? 
+    - If it will be the same for all instances, not intended for the caller to fill in, and only accessed via `Unwrap`, we can actually hide the field.
+    - The wrapped errors seem like the super class. It may be more reasonable to wrap other errors instead of being wrapped. But the order does not matter in the error chain.
+
+## Proposal
+
+So I propose the following error handling specification as a supplement of `11-error-handling`:
+
+1. Public APIs SHOULD return top-level errors, which MUST be defined as below:
```

Comment:
Please make it clear that what's the top-level errors look like. Will we keep the `InitError`, `ServiceError` and `StorageError`?